Deferrable sensors can implement sensor timeout

dstandish opened the pull request:
The goal here is to allow behavioral parity w.r.t. sensor timeouts between deferrable and non-deferrable sensor operators.

With non-deferrable sensors, if there's a sensor timeout, the task fails without retry.  But currently, with deferrable sensors, that does not happen.

Since there's already a "timeout" capability on triggers, we can use this for sensor timeout.  Essentially all that was missing was the ability to distinguish between trigger timeouts and other trigger errors.  With this capability, base sensor can distinguish between the two, and reraise deferral timeouts as sensor timeouts.

So, here we add a new exception type, TaskDeferralTimeout, which base sensor reraises as AirflowSensorTimeout. Then, to take advantage of this feature, a sensor need only ensure that its timeout is passed when deferring. For convenience, we update the task deferred exception signature to take int and float in addition to timedelta, since that's how `timeout` attr is defined on base sensor.  But we do not change the exception attribute type.

In order to keep this PR focused, this PR only updates one sensor to use the timeout functionality, namely, time delta sensor.  Other sensors will have to be done as followups.


-------

Old description below ⬇️ 

Alternative to https://github.com/apache/airflow/pull/32990

resolves https://github.com/apache/airflow/issues/32638

This is a less invasive approach.  Essentially, what we do here is, update `BaseOperator.resume_execution` so that when trigger times out then it raises special exception `AirflowDeferralTimeout`.

Then, `BaseSensorOperator.resume_execution`, we reraise `AirflowDeferralTimeout` as a `AirflowSensorTimeout`.

So, if a sensor resumes from a timed-out deferral, then it's interpreted as a sensor timeout.

All that is required is for a sensor to add a timeout to the deferral.

Example logs:

```
...
[2023-08-25, 07:00:54 UTC] {taskinstance.py:1357} INFO - Resuming after deferral
[2023-08-25, 07:00:54 UTC] {taskinstance.py:1380} INFO - Executing <Task(TimeDeltaSensorAsync): delta_sensor> on 2023-08-25 07:00:32.864128+00:00
[2023-08-25, 07:00:54 UTC] {standard_task_runner.py:57} INFO - Started process 50543 to run task
[2023-08-25, 07:00:54 UTC] {standard_task_runner.py:84} INFO - Running: ['airflow', 'tasks', 'run', 'simple2', 'delta_sensor', 'manual__2023-08-25T07:00:32.864128+00:00', '--job-id', '129', '--raw', '--subdir', '/Users/dstandish/code/airflow/airflow/example_dags/async_timeout.py', '--cfg-path', '/var/folders/9c/tknx7xx10qx92983y1r5djb40000gn/T/tmp72slvjz_']
[2023-08-25, 07:00:54 UTC] {standard_task_runner.py:85} INFO - Job 129: Subtask delta_sensor
[2023-08-25, 07:00:54 UTC] {task_command.py:415} INFO - Running <TaskInstance: simple2.delta_sensor manual__2023-08-25T07:00:32.864128+00:00 [running]> on host daniels-mbp.lan
[2023-08-25, 07:00:54 UTC] {taskinstance.py:1933} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/Users/dstandish/code/airflow/airflow/sensors/base.py", line 288, in resume_execution
    return super().resume_execution(next_method, next_kwargs, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstandish/code/airflow/airflow/models/baseoperator.py", line 1605, in resume_execution
    raise AirflowDeferralTimeout(error)
airflow.exceptions.AirflowDeferralTimeout: Trigger timeout
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/Users/dstandish/code/airflow/airflow/sensors/base.py", line 290, in resume_execution
    raise AirflowSensorTimeout(*e.args) from e
airflow.exceptions.AirflowSensorTimeout: Trigger timeout
[2023-08-25, 07:00:54 UTC] {taskinstance.py:1398} INFO - Immediate failure requested. Marking task as FAILED. dag_id=simple2, task_id=delta_sensor, execution_date=20230825T070032, start_date=20230825T070034, end_date=20230825T070054
[2023-08-25, 07:00:54 UTC] {standard_task_runner.py:104} ERROR - Failed to execute job 129 for task delta_sensor (Trigger timeout; 50543)
[2023-08-25, 07:00:54 UTC] {local_task_job_runner.py:228} INFO - Task exited with return code 1
[2023-08-25, 07:00:54 UTC] {taskinstance.py:2774} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2023-08-25, 07:00:55 UTC] {triggerer_job_runner.py:611} ERROR - Trigger cancelled due to timeout
[2023-08-25, 07:00:55 UTC] {triggerer_job_runner.py:612} ERROR - Trigger cancelled; message=
```

dstandish commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1675,7 +1675,7 @@ def check_trigger_timeouts(self, session: Session = NEW_SESSION) -> None:
                 state=TaskInstanceState.SCHEDULED,
                 next_method="__fail__",
-                next_kwargs={"error": "Trigger/execution timeout"},
+                next_kwargs={"error": "Trigger timeout"},
```
this i change just to be clear the scenario is exclusively trigger timeout, not execution_timeout, IIUC

dstandish commented on airflow/exceptions.py:
```diff
@@ -366,6 +370,8 @@ def __init__(
         self.kwargs = kwargs
         self.timeout = timeout
         # Check timeout type at runtime
+        if isinstance(self.timeout, (int, float)):
```
this is just for convenience since BaseSensorOperator coalesces timeout from timedelta to seconds.

dstandish commented on airflow/sensors/time_delta.py:
```diff
@@ -64,7 +64,11 @@ class TimeDeltaSensorAsync(TimeDeltaSensor):
+        self.defer(
+            trigger=DateTimeTrigger(moment=target_dttm),
+            method_name="execute_complete",
+            timeout=self.timeout,
```
this adds a timeout to the deferral for this sensor.  and as a result, if the deferral times out, then the task will fail without retry

hussein-awala commented on airflow/sensors/time_delta.py:
```diff
@@ -64,7 +64,11 @@ class TimeDeltaSensorAsync(TimeDeltaSensor):
+        self.defer(
+            trigger=DateTimeTrigger(moment=target_dttm),
+            method_name="execute_complete",
+            timeout=self.timeout,
```
I think in this case, we'll need to explicitly add `timeout=min(self.timeout, custom_timeout)` in all the sensors? (just self.timeout if we don't want to add a custom timeout for the trigger)

dstandish replied:
what do you mean custom_timeout?  what is that?  There need not be any custom timeout?  I think we would only have to add `timeout=self.timeout` to all sensors.

dstandish replied:
then it's simply the behavior that, for BaseSensorOperator, the deferral timeout is used for `timeout`

hussein-awala replied:
From the other PR:
> it makes me a bit uncomfortable to remove the feature to be able to set a specific deferral timeout on a given deferral. not that it's necessarily all that useful, but it's there.

So for the sensors, this deferral timeout should be always self.timeout and added explicitly.

Also if the sensor fails for some reason, we will reset this timeout. For example a sensor with 5 retries and 1h as timeout, if it fails after 30min on each retry because a token is expired, the sensor will run for 2h30min > 1h

dstandish replied:
> it makes me a bit uncomfortable to remove the feature to be able to set a specific deferral timeout on a given deferral. not that it's necessarily all that useful, but it's there.

That comment I think was in reference to an earlier version of your PR, when it was deprecating the deferral timeout altogether.

dstandish replied:
> Also if the sensor fails for some reason, we will reset this timeout. For example a sensor with 5 retries and 1h as timeout, if it fails after 30min on each retry because a token is expired, the sensor will run for 2h30min > 1h

Can you clarify which scenario you are talking about.

Maybe just clarify, with this example, what is the desired behavior that we need to design for, and how this design perhaps falls short, in that it cannot be used to produce the desired behavior, or is inconsistent.

dstandish replied:
Because.... like there's lots of different scenarios floating around: this solution, your solution, the non-async version, and current state.  So just hoping to understand clearly where things diverge and how they do so we can figure out what's best.

hussein-awala replied:
> Can you clarify which scenario you are talking about.

> Maybe just clarify, with this example, what is the desired behavior that we need to design for, and how this design perhaps falls short.

When we add `timeout=1h` for a sensor, we expect that the total execution date for this sensor is 1h, even if fails and retires multiple time.

From the doc:
> timeout ([datetime.timedelta](https://docs.python.org/3/library/datetime.html#datetime.timedelta) | [float](https://docs.python.org/3/library/functions.html#float)) – Time elapsed before the task times out and fails. Can be timedelta or float seconds. This should not be confused with execution_timeout of the BaseOperator class. timeout measures the time elapsed between the first poke and the current time (taking into account any reschedule delay between each poke), while execution_timeout checks the running time of the task (leaving out any reschedule delay). In case that the mode is poke (see below), both of them are equivalent (as the sensor is never rescheduled), which is not the case in reschedule mode.

So currently, in reschedule mode, the duration is calculated as:
```python
run_duration = (timezone.utcnow() - start_date).total_seconds()

# and start_date is
first_try_number = context["ti"].max_tries - self.retries + 1
task_reschedules = TaskReschedule.find_for_task_instance(
    context["ti"], try_number=first_try_number
)
if not task_reschedules:
    start_date = timezone.utcnow()
else:
    start_date = task_reschedules[0].start_date
```
Personally, I have a Glue sensor, and I'm setting `timeout` to 4h (it waits for a partition in a table). Sometimes the poke method fails because there is a pressure on the Glue API, for that I set retries=10, and I don't care even if it's 100, because the sensor will timeout after 4h regardless the number of attempts.

This behavior could be unexpected in Airflow, but it's the current behavior in the sync mode, and the users (including me) use it like that. So if we want to change something, at least we need to update the documentation and explain why this is different, and maybe fix the sync mode if there is an unexpected behavior.

dstandish replied:
OK so it sounds like you're saying that the problem with the approach in the PR currently is that, after a retry, when we defer again, we don't adjust the _defer_ timeout to be relative to the original start_date.

And it sounds like what we would need to do is, timeout that we pass to TaskDeferred must be calculated as `(orig_start_date + self.timeout) - current_time`.  And then, the sensor timeout will be evaluated relative to the first attempt, and not from the start date of a retry.  And then perhaps this would make the behavior consistent?

hussein-awala commented on airflow/sensors/base.py:
```diff
@@ -285,6 +286,8 @@ def run_duration() -> float:
     def resume_execution(self, next_method: str, next_kwargs: dict[str, Any] | None, context: Context):
         try:
             return super().resume_execution(next_method, next_kwargs, context)
+        except TaskDeferralTimeout as e:
+            raise AirflowSensorTimeout(*e.args) from e
```
With this change, you stop the sensor execution when timeout is reached regardless its reason. Is it expected? Because if so, sync and async sensor will be inconsistent.

dstandish replied:
What this means is that for sensors, deferral timeout is used for sensor timeout and it just means one thing -- it means there is no other reason but sensor timeout.

dstandish replied:
What inconsistency do you see?

hussein-awala replied:
Sensor is an operator, so it accept execution_timeout as param, and currently in sync mode, when we provide:
retries=2
execution_timeout=1h
retry_delay=30m
timeout=3h30

The sensor will runs for 1h, timeout, retries after 30m, runs for 1h, timeout, retries after 30m and timeout after 30m with failed state because the total time is 3h30.

> deferral timeout is used for sensor timeout and it just means one thing -- it means there is no other reason but sensor timeout.

In your case you fail sensor when one of the execution timeout or the timeout is reached. Is allowing providing execution_time in base_sensor is a bug for you?

If not, we will have a difference between handling these two configs between sync and async modes, and that is what I mean by inconsistent.

dstandish replied:
Sorry, with your scenario, can you clarify what will be the difference in behavior between async and sync?

But aside from this, perhaps it would be helpful to articulate the behavior we want to design for, and ask the question, does this design enable this behavior.  I am not sure a perfect 1-1 consistency in all parameter combinations is really necessary; it may not even be practical, or desirable in some cases.  E.g. some parameter combinations may be nonsensical or edge case manifestations, and it may not be worth the trouble to go out of our way to reproduce them in the deferrable case.

But so for example with this one:

> The sensor will runs for 1h, timeout, retries after 30m, runs for 1h, timeout, retries after 30m and timeout after 30m with failed state because the total time is 3h30.

This seems like a misconfiguration.  This behavior seems essentially equivalent to having no execution_timeout at all.

Concerning what behavior we actually want to design for, the problem as I understood it is that we want to be able to use `timeout` to signal to a sensor to fail with no more retries.  This does that.  What else is missing?

dstandish replied:
@hussein-awala shall we pick this convo back up?  Let's figure out if this PR is good to merge or needs changes or just needs to be closed.

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1600,7 +1601,10 @@ def resume_execution(self, next_method: str, next_kwargs: dict[str, Any] | None,
             if traceback is not None:
                 self.log.error("Trigger failed:\n%s", "\n".join(traceback))
-            raise TaskDeferralError(next_kwargs.get("error", "Unknown"))
+            if (error := next_kwargs.get("error", "Unknown")) == "Trigger timeout":
```
I wonder if this can be made a bit more robust… Would an additional key in next_kwargs be possible?

dstandish replied:
do you have any suggestion in particular?
like....
```
{"__internal__": "trigger_timeout"}
```
or something?

dstandish replied:
@uranusjr any particular suggestion you have?

dstandish committed:
Deferrable sensors can implement sensor timeout

The goal here is to ensure behavioral parity w.r.t. sensor timeouts between deferrable and non-deferrable sensor operators.

With non-deferrable sensors, if there's a sensor timeout, the task fails without retry.  But currently, with deferrable sensors, that does not happen.

Since there's already a "timeout" capability on triggers, we can use this for sensor timeout.  Essentially all that was missing was the ability to distinguish between trigger timeouts and other trigger errors.  With this capability, base sensor can distinguish between the two, and reraise deferral timeouts as sensor timeouts.

So, here we add a new exception type, TaskDeferralTimeout, which base sensor reraises as AirflowSensorTimeout. Then, to take advantage of this feature, a sensor need only ensure that its timeout is passed when deferring. For convenience, we update the task deferred exception signature to take int and float in addition to timedelta, since that's how `timeout` attr is defined on base sensor.  But we do not change the exception attribute type.

In order to keep this PR focused, this PR only updates one sensor to use the timeout functionality, namely, time delta sensor.  Other sensors will have to be done as followups.

dstandish committed:
fix typing

dstandish committed:
use constant

dstandish committed:
use enum for trigger failure reasons

dstandish committed:
fix test

dstandish committed:
fix import

uranusjr commented on airflow/exceptions.py:
```diff
@@ -417,6 +421,10 @@ class TaskDeferralError(AirflowException):
     """Raised when a task failed during deferral for some reason."""
 
 
+class TaskDeferralTimeout(AirflowException):
```
Should this subclass TaskDeferralError for compatibility? Or do we specifically want the two errors to be distinct?

dstandish replied:
I think it's fine to subclass.  I will change it.

dstandish replied:
well actually, let's think about this

dstandish replied:
So, TaskDeferralError is raised when something unexpected goes wrong on the trigger side of things.

This class, TaskDeferralTimeout, is actually initiated by the scheduler, here

<img width="765" alt="image" src="https://github.com/user-attachments/assets/643ea8d9-6b3b-412a-8973-08b17cbb9e9d">

So, it might actually make sense as two distinct families.  I guess it depends how you think about it.  What do you think @uranusjr ?

eladkal commented on providers/src/airflow/providers/standard/sensors/time_delta.py:
```diff
@@ -91,7 +91,11 @@ def execute(self, context: Context) -> bool | NoReturn:
+            trigger=trigger,
+            method_name="execute_complete",
+            timeout=self.timeout,
+        )
```
Just making sure - is this change compatible with Airflow >= 2.8?

dstandish replied:
hrm... no we will need to add some backcompat, i think.  in which airflow version do you think this should be released?  probably 2.11. i guess that's what i'll assume.

dstandish replied:
ok @eladkal 

i have added the backcompat logic. i will resolve this convo, but please review

dstandish committed:
add backcompat logic

eladkal reviewed (approved):
LGTM

But we must make sure the core part of this PR is released in 2.11 otherwise we will have compatibility problems with the provider part cc @ephraimbuddy @utkarsharma2

dstandish committed:
simplify backcompat

dstandish committed:
fixup

dstandish committed:
fixup
